esp/services/esdl_svc_engine/esdl_monitor.cpp

```diff
@@ -0,0 +1,576 @@
+    {
+        if (!defId || !*defId)
+            return;
+        StringBuffer lcid(defId);
+        lcid.trim().toLowerCase();
+        if (lcid.length() == 0)
+            return;
+        defId = lcid.str();
+        DBGLOG("Removing esdl definition %s from shared cache", defId);
+        {
+            CriticalBlock cb(m_critSect);
+            m_esdlInstanceMap.remove(defId);
+        }
+    }
+
+    virtual Linked<IEsdlDefinition> query(const char* defId) override
+    {
+        if (!defId || !*defId)
+            return nullptr;
+        StringBuffer lcid(defId);
+        lcid.trim().toLowerCase();
+        if (lcid.length() == 0)
+            return nullptr;
+        defId = lcid.str();
+        {
+            CriticalBlock cb(m_critSect);
+            Owned<CEsdlInstance>* instancepp = m_esdlInstanceMap.getValue(defId);
+            if (instancepp)
+                return (*instancepp)->m_def.get();
+        }
+        return nullptr;
+    }
+
+    //Thread periodically cleaning definition objects that are no longer being used
+    virtual int run() override
+    {
+        while (!m_stop)
+        {
+            m_waitsem.wait(300000);
+            if(m_stop)
+                break;
+            {
+                CriticalBlock cb(m_critSect);
+                StringArray ids;
+                for (auto& item : m_esdlInstanceMap)
+                {
+                    CEsdlInstance* instance = item.getValue();
+                    if (!instance || !instance->m_def.get())
+                        continue;
+                    if (!instance->m_def->isShared())
+                        ids.append(instance->m_defId.str());
+                }
+                ForEachItemIn (x, ids)
+                {
+                    const char* id = ids.item(x);
+                    DBGLOG("Definition %s is no longer being used, remove it from esdl shared cache", id);
+                    m_esdlInstanceMap.remove(id);
+                }
+            }
+        }
+
+        return 0;
+    }
+
+    virtual void stop()
+    {
+        m_stop = true;
+        m_waitsem.signal();
+    }
+private:
+    CriticalSection m_critSect;
+    MapStringTo<Owned<CEsdlInstance>> m_esdlInstanceMap;
+    Semaphore m_waitsem;
+    bool m_stop;
+};
+
+static CriticalSection gEsdlMonitorCritSection;
+
+class CEsdlMonitor : implements IEsdlMonitor, public CInterface, implements IEsdlListener
+{
+private:
+    MapStringTo<EsdlBindingImpl*> m_esdlBindingMap;
+    CriticalSection m_CritSect;
+    StringBuffer mEnvptHeader;
+    Owned<IEsdlStore> m_pCentralStore;
+    Owned<IEsdlSubscription> m_pSubscription;
+    Owned<CEsdlShare> m_esdlShare;
+    bool m_isSubscribed;
+
+public:
+    IMPLEMENT_IINTERFACE;
+
+    CEsdlMonitor() : m_isSubscribed(false)
+    {
+        constructEnvptHeader();
+        m_pCentralStore.setown(createEsdlCentralStore());
+        m_esdlShare.setown(new CEsdlShare());
+        m_esdlShare->start();
```

Comment:
What stops and/or joins this thread?

Reply:
The stopEsdlMonitor() method which will be called by ESP's shutdown mechanism will stop this thread.

Reply:
ok I see, it's quite confusing. It only works because there's 1 instance of CEsdlMonitor really.

Also, stopEsdlMonitor() calls stop() but doesn't join the thread, which could lead to problems.

It would be better to contain the thread in the owning class and have a EsdlMonitor::stop() method, stop and join it's own thread.

Reply:
I've added the join to the stop method. Good point to add the join, but I think logically it makes more sense to keep the thread in the share class instead of the monitor class.